lib/catalog/owid/catalog/tables.py

```diff
@@ -530,3 +578,638 @@ def join(self, other: Union[pd.DataFrame, "Table"], *args, **kwargs) -> "Table":
+        # TODO: Arguments left_index/right_index are not implemented.
+        raise NotImplementedError(
+            "Arguments 'left_index' and 'right_index' currently not implemented in function 'merge'."
+        )
+    # Create merged table.
+    tb = Table(
+        pd.merge(
+            left=left, right=right, how=how, on=on, left_on=left_on, right_on=right_on, suffixes=suffixes, **kwargs
+        )
+    )
+
+    # If arguments "on", "left_on", or "right_on" are given as strings, convert them to lists.
+    if isinstance(on, str):
+        on = [on]
+    if isinstance(left_on, str):
+        left_on = [left_on]
+    if isinstance(right_on, str):
+        right_on = [right_on]
+
+    if (on is None) and (left_on is None):
+        # By construction, either "on" is passed, or both "left_on" and "right_on".
+        # Any other possibility will raise a MergeError, and hence doesn't need to be considered.
+        # If none of them is not specified, assume we are joining on common columns.
+        on = list(set(left.all_columns) & set(right.all_columns))
+        left_on = []
+        right_on = []
+    elif left_on is not None:
+        # By construction, right_on must also be given.
+        on = []
+    else:
+        # Here, "on" is given, but not "left_on".
+        left_on = []
+        right_on = []
+
+    # Find columns that existed in both left and right tables whose name will be modified with suffixes.
+    overlapping_columns = ((set(left.all_columns) - set(left_on)) & (set(right.all_columns) - set(right_on))) - set(on)  # type: ignore
+
+    # Find columns that existed in both left and right tables that will preserve their names (since they are columns to join on).
+    common_columns = on or (set(left_on) & set(right_on))  # type: ignore
+
+    columns_from_left = set(left.all_columns) - set(common_columns)
+    columns_from_right = set(right.all_columns) - set(common_columns)
+
+    for column in columns_from_left:
+        if column in overlapping_columns:
+            new_column = f"{column}{suffixes[0]}"
+        else:
+            new_column = column
+        tb[new_column].metadata = variables.combine_variables_metadata([left[column]], operation="merge", name=column)
+
+    for column in columns_from_right:
+        if column in overlapping_columns:
+            new_column = f"{column}{suffixes[1]}"
+        else:
+            new_column = column
+        tb[new_column].metadata = variables.combine_variables_metadata([right[column]], operation="merge", name=column)
+
+    for column in common_columns:
+        tb[column].metadata = variables.combine_variables_metadata(
+            [left[column], right[column]], operation="merge", name=column
+        )
+
+    # Update table metadata.
+    tb.metadata.title = combine_tables_titles(tables=[left, right])
+    tb.metadata.description = combine_tables_descriptions(tables=[left, right])
+    tb.metadata.dataset = DatasetMeta(
```

Comment:
_This is probably not a blocker for merging, but definitely an issue we need to address at some point_

`TableMeta.dataset` is used to reference dataset metadata. The problem is that the reference doesn't point to the same object and `dataset.metadata` and `tb.metadata.dataset` is different. I'm not sure why it was implemented like this, but I've already run into this inconsistency somewhere else, thinking that if I do `tb.metadata.dataset.description = "foo"`, it would change `ds.metadata.description`.

So what you're doing here is replacing the reference with the new `DatasetMeta` object and `tb.metadata.dataset` isn't a reference anymore (and both are out-of-sync and might be re-synced elsewhere).

I don't know how it affects your PR yet, will have to dig deeper. Anyway, it'd be worth raising this inconsistency as a separate issue and think about what do we want out of it (and possible forbid setting `tb.metadata.dataset`). @larsyencken mind sharing your thoughts?

Here's an example demonstrating that metadata are different objects
```
from owid import catalog
from etl.paths import DATA_DIR

ds = catalog.Dataset(DATA_DIR / "garden/regions/2023-01-01/regions")
id(ds['regions'].metadata.dataset), id(ds.metadata)
```

Reply:
Currently there's no other way to store sources at the table level. I don't think it would be better to refer to the dataset of any of the input tables. Solving this properly would imply changing the attributes of datasets and tables (which is something to be handled in another PR). So, do you think there's anything else to do for this PR?

Reply:
The reason I'm worried is that `tb.metadata.dataset` is used when upserting data to MySQL (which we haven't tested much). By assigning to `tb.metadata.dataset` you're replacing the reference with an object you're using to propagate sources and licenses. I'm not saying it'll necessarily break things, but it's a technical debt that will have to be dealt with sooner or later. (I'm cautious here because I spent 2 hours last week debugging a problem caused by `tb.metadata.dataset` and I'm worried about introducing technical debt before redoing our metadata structure).

Since the sources of a dataset are the union of indicators sources, would it be possible to do this at the very end when calling `create_dataset` instead of doing it with every operation? (Ideally, `dataset.sources` would be calculated dynamically, but we don't have to do it as part of this)

(After I finish `origins`, I'll start looking into `tb.metadata.dataset` and try to make it read-only property)

Reply:
The reason it's implemented this way was that we chose _tables_ as the core unit for the ETL, since we were trying to make an ideal data science environment and that's how people typically do data analysis.

That means the format needs to let you load a table and get everything you need. For that reason, the dataset-level metadata is duplicated to the table any time you add a table to the dataset: https://github.com/owid/etl/blob/0dffc36024c625b35d872e0afc22927520bcaf86/lib/catalog/owid/catalog/datasets.py#L121

Since it is only cached at the table level, we should definitely not write to the table's copy directly. The correct/best way to do this is to store this metadata at the variable level instead.